ECO-936: Casper glossary

ipopescu opened the pull request:
Added a glossary to the docs based on this content:
https://casperlabs.atlassian.net/wiki/spaces/EN/pages/26280411/CasperLabs+Glossary+of+Terms+and+Spelling+Guide
Added some additional terms in commit f358a43.
- The glossary is in the main top navigation.
- The "on this page" section lists all the terms
- Each term is a header to be referenced as a link in another glossary explanation.
- Each glossary page has links to all the letters for navigation
- Some sections like section Q, are empty saying "coming soon"

<img width="1113" alt="Screenshot 2021-02-22 at 16 51 02" src="https://user-images.githubusercontent.com/4185994/108733447-ce98d500-752e-11eb-87c4-14e55765624b.png">

<img width="933" alt="Screenshot 2021-02-22 at 16 51 17" src="https://user-images.githubusercontent.com/4185994/108733552-e3756880-752e-11eb-97d0-6a4c19a5700a.png">

<img width="895" alt="Screenshot 2021-02-22 at 16 51 29" src="https://user-images.githubusercontent.com/4185994/108734008-4d8e0d80-752f-11eb-8f7b-0ec2feec72dd.png">

<img width="673" alt="Screenshot 2021-02-22 at 16 51 46" src="https://user-images.githubusercontent.com/4185994/108733683-056eeb00-752f-11eb-8e2e-fcb2f9d318a2.png">

<img width="689" alt="Screenshot 2021-02-22 at 16 51 55" src="https://user-images.githubusercontent.com/4185994/108733816-1b7cab80-752f-11eb-85b1-5cf1b92e26cd.png">

MParlikar commented on glossary/S.rst:
```diff
@@ -0,0 +1,34 @@
+
+Seigniorage
+^^^^^^^^^^^
+The reward mechanism by which we mint new tokens and distribute them to validators.
```
'The reward mechanism by which validators are rewarded for participating in consensus.  New tokens are minted and given to validators'

ipopescu replied:
Updated

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+You will encounter different types of nodes on the network:
+
+* **Bootstrap node**: a peer node on the P2P network that contains the logic for message passing between nodes on the network (via TCP) and discovering new peers so that they may join the network.
+* **Stand-alone node**: a mode for starting the node in which it does not attempt to connect to another node that is already part of a P2P network (called the bootstrap node). This mode is useful for local testing.
```
I do not believe this is supported.   A stand alone Casper node will panic.

ipopescu replied:
Removed the standalone node; however, I kept the bootstrap and read-only nodes.

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+Nodes are constructed of several software components or layers: Client, COMM, Consensus, Execution Engine, and Storage. 
+
+#. gRPC Client - gRPC interfaces to various node functions, including accepting deploys, creating+proposing `blocks <B.html#block>`_ (not exposed publicly), and querying the `blockDAG <B.html#blockdag>`_
+#. COMM - contains the logic for message passing between nodes on the network (via TCP) as well as discovering new peers (via Kademlia)
```
We do not have Kademlia node discover right now.

ipopescu replied:
Removed references to Kademlia and fragments containing it such as "discovering new peers (via Kademlia)".

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+
+#. gRPC Client - gRPC interfaces to various node functions, including accepting deploys, creating+proposing `blocks <B.html#block>`_ (not exposed publicly), and querying the `blockDAG <B.html#blockdag>`_
+#. COMM - contains the logic for message passing between nodes on the network (via TCP) as well as discovering new peers (via Kademlia)
+#. Consensus - contains the logic to update the `blockstore <B.html#blockstore>`_ so that agreement is maintained between nodes. It also contains logic to detect when `deploys <D.html#deploy>`_ commute (this enables the use of parallelism in block creation and in consensus itself via the blockDAG)
```
block store and global state.  We do not support commutation (paralelism) at this time

ipopescu replied:
Added global state. Removed the sentence about commutation.

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+
+You will encounter different types of nodes on the network:
+
+* **Bootstrap node**: a peer node on the P2P network that contains the logic for message passing between nodes on the network (via TCP) and discovering new peers so that they may join the network.
```
We call these 'trusted nodes' I believe in config.toml.  It could be any node on the network that can be trusted.

ipopescu replied:
Changed to "trusted node" instead of "bootstrap node".

ipopescu replied:
Given the later comment, changed to bonded node.

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+
+* **Bootstrap node**: a peer node on the P2P network that contains the logic for message passing between nodes on the network (via TCP) and discovering new peers so that they may join the network.
+* **Stand-alone node**: a mode for starting the node in which it does not attempt to connect to another node that is already part of a P2P network (called the bootstrap node). This mode is useful for local testing.
+* **Read-only node**: a type of node on the network that receives and processes blocks but does not create blocks and is not a validator. It is otherwise a fully functioning node, following the consensus protocol to know the current status of the blockDAG (and therefore also the VM state). Such nodes are useful for querying the status of the blockDAG (e.g., to learn information about transaction  ...
```
We should change this to 'Bonded and Unbonded nodes' - All nodes have to start up the same way.  The difference is whether the node becomes bonded  through a separate bid to the auction contract.  In this way, we are pretty different!

ipopescu replied:
Great, updated to use bonded/unbonded nodes.

MParlikar commented on glossary/N.rst:
```diff
@@ -0,0 +1,33 @@
+
+Node bootstrap
+^^^^^^^^^^^^^^
+In `node <N.html#node>`_ COMM - contains the logic for message passing between nodes on the network (via TCP) as well as discovering new peers (via Kademlia).
```
We call these 'trusted nodes'.  Feel free to add the definition here.

ipopescu replied:
I moved and updated this entry under T for _trusted node_: 
A bonded node that contains the logic for message passing between nodes on the network (via TCP).

MParlikar commented on glossary/P.rst:
```diff
@@ -0,0 +1,27 @@
+
+Participate in consensus
+^^^^^^^^^^^^^^^^^^^^^^^^
+The process of following the `consensus <C.html#consensus>`_ algorithm. The primary participants are `validators <V.html#validators>`_.
```
Please also include 'delegators' for Casper consensus. Delegators (stakers) add weight to validators, and can be the deciding factor on who is in or out..

ipopescu replied:
Added entries for "delegator" and "delegation rate".

**Delegation rate**
Node operators (validators) define a commission that they take in exchange for providing staking services. This commission is represented as a percentage of the rewards that the node operator retains for their services.

**Delegator**
Delegators are users who participate in the platform's security by delegating their tokens to validators, which adds to their weight, and collecting a part of the rewards proportional to their delegations, net of a cut ("delegation rate") that is collected by the validator.

MParlikar commented on glossary/P.rst:
```diff
@@ -0,0 +1,27 @@
+Participate in consensus
+^^^^^^^^^^^^^^^^^^^^^^^^
+The process of following the `consensus <C.html#consensus>`_ algorithm. The primary participants are `validators <V.html#validators>`_.
+
```
Casper also has the notion of payment code separate from Session code.

ipopescu replied:
Added a new entry for payment code: the WASM program that pays the transaction execution fee.

MParlikar commented on glossary/S.rst:
```diff
@@ -0,0 +1,34 @@
+
+Staker
+^^^^^^
+A person that deposits tokens in the `proof-of-stake <P.html#proof-of-stake>`_ contract. Sometimes a staker is also a `validator <V.html#validator>`_. Stakers take on the slashing risk in exchange for rewards. Stakers will deposit their `tokens <T.html#token>`_ by sending a bonding request in the form of a transaction (deployment) to the system. If a validator is `slashed <#slashing>`_, the stake ...
```
A staker can also be a delegator

ipopescu replied:
Updated to say that a staker can be a validator or delegator.

MParlikar commented on glossary/P.rst:
```diff
@@ -0,0 +1,27 @@
+Primary token
+^^^^^^^^^^^^^
+The Casper platform defines only one token which one can pay for transaction execution. Presently this is called `CSPR <C.html#cspr>`_.
+
```
Include private key  (we like to call them secret keys in our docs)

ipopescu replied:
I added 2 entries for _private key_, and _secret key_, with this definition: "A cryptographic and confidential key that signs transactions to ensure a valid execution can only do the user's intended operations."

MParlikar commented on glossary/T.rst:
```diff
@@ -0,0 +1,44 @@
+*  **Nonce** – [mandatory] the number which prevents replay attacks by making each signature valid for a single execution
+*  **TTL** - [optional] upper limit for P-time of a block this transaction may be included in
+*  **Block hook** - [optional] semantic dependency of this transaction, expressed as a hash of a block that is 
+*  **Signature** – [mandatory] must be over (payment code, session code, conversion rate, nonce) to ensure a valid execution can only do the operations the user wants to be done.
```
Include that the signature is done via a private or secret key

ipopescu replied:
Added

MParlikar commented on glossary/U.rst:
```diff
@@ -0,0 +1,12 @@
+
+Unbonding
+^^^^^^^^^
+Withdrawing money from the Proof-of-Stake (PoS) contract and possibly ceasing to be a validator.
```
Auction contract.  Done by 'withdraw bid'

ipopescu replied:
Updated

MParlikar commented on glossary/U.rst:
```diff
@@ -0,0 +1,12 @@
+Unbonding
+^^^^^^^^^
+Withdrawing money from the Proof-of-Stake (PoS) contract and possibly ceasing to be a validator.
+The unbonding request is a transaction that informs the PoS contract that the sender wants to decrease their deposit. In the next booking block, only the decreased deposit is considered when determining a future validator set. If it has been decreased to 0, the sender will not be included in the validator set anymore.
```
Auction contract.

ipopescu replied:
Updated

MParlikar commented on glossary/U.rst:
```diff
@@ -0,0 +1,12 @@
+^^^^^^^^^
+Withdrawing money from the Proof-of-Stake (PoS) contract and possibly ceasing to be a validator.
+The unbonding request is a transaction that informs the PoS contract that the sender wants to decrease their deposit. In the next booking block, only the decreased deposit is considered when determining a future validator set. If it has been decreased to 0, the sender will not be included in the validator set anymore.
+However, the amount only gets transferred to the sender several weeks later, after the unbonding period. If during that period their node exhibits a fault, the unbonded amount can still be slashed.
```
remove 'several weeks later'

ipopescu replied:
Done

MParlikar commented on glossary/V.rst:
```diff
@@ -0,0 +1,17 @@
+* `block validation <B.html#block-validation>`_
+* `block gossiping <B.html#block-gossiping>`_
+
+Validators are bonded because they are responsible for progressing the system's state as clients use it (e.g., sending deploys). Validators and `stakers <S.html#staker>`_ can lose their bond (be slashed) for not following the protocol correctly. Validators are also paid by creating blocks (also by validating blocks – though this is only indirectly; validators cannot be paid if they do not validat ...
```
paid 'for'

ipopescu replied:
Updated

MParlikar reviewed (changes requested):
Great job.  Much needed.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,95 @@
+
+Bonding
+^^^^^^^
+Depositing money in the Proof-of-Stake (PoS) contract and trying to become a `staker <S.html#staker>`_. The bonding request is a transaction that transfers tokens to the PoS contract. In the next `booking block <B.html#booking-block>`_, a new set of validators is determined, with weights according to their deposits. This new set becomes active in the era(s) using that booking block.
```
Updated to: "Depositing money in the auction contract and try to become a staker. The bonding request is a transaction that transfers tokens to the auction contract. In the next booking block, a new set of validators is determined, with weights according to their deposits. This new set becomes active in the era(s) using that booking block."

ipopescu commented on glossary/A.rst:
```diff
@@ -0,0 +1,10 @@
+
+AssemblyScript
+^^^^^^^^^^^^^^
+Is a TypeScript-based programming language (essentially JavaScript with static types) that is optimized for WebAssembly and compiled to WebAssembly using asc, the reference AssemblyScript compiler. It is developed by the AssemblyScript Project and the AssemblyScript community.
```
Added a new term: 
**Auction**
The auction determines the composition of the validator set for each era of the protocol. It is a "first-price" auction (where winning bids become stakes) with a fixed number of spots, chosen to balance security with performance. Because rewards are proportional to the stake, it is expected that this competitive mechanism will provide a powerful impetus for staking as many tokens as possible.

ipopescu commented on glossary/P.rst:
```diff
@@ -0,0 +1,27 @@
+Proof of work
+^^^^^^^^^^^^^
+A mechanism used in Bitcoin and Etherium for incentivizing participation and securing the system. In these protocols, the voting power of a participant is proportional to the amount of computational power possessed.
+
```
Yes, added proto block.

ipopescu commented:
Ready for a second review.
I addressed the feedback, updated the Validator entry, and added new entries for:
- Ecosystem
- Eviction
- Proof-of-Stake contract
- Rewards
- Runtime
- Transaction fees
- Users

afck commented on glossary/index.rst:
```diff
@@ -38,3 +34,10 @@ These definitions are correct in the context of the CasperLabs. They may (and pr
    X.rst
    Y.rst
    Z.rst
+<<<<<<< HEAD
+=======
```
Looks like a merge conflict?
```suggestion
```

ipopescu replied:
Yes, resolved, sorry about that!

afck commented on glossary/index.rst:
```diff
@@ -38,3 +34,10 @@ These definitions are correct in the context of the CasperLabs. They may (and pr
+============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== 
+`A <A.html>`_  `B <B.html>`_  `C <C.html>`_  `D <D.html>`_  `E <E.html>`_  `F <F.html>`_  `G <G.html>`_  `H <H.html>`_  `I <I.html>`_  `J <J.html>`_  `K <K.html>`_  `L <L.html>`_  `M <M.html>`_  `N <N.html>`_  `O <O.html>`_  `P <P.html>`_  `Q <Q.html>`_  `R <R.html>`_  `S <S.html>`_  `T <T.html>`_  `U <U.html>`_  `V <V.html>`_  `W <W.html>`_  `X <X.html>`_  `Y <Y.html>`_  `Z <Z.html>`_  
+============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== 
+>>>>>>> 1829d72 (initial commit)
```
```suggestion
```
(Or is the whole paragraph supposed to be removed?)

ipopescu replied:
Somehow this got through without raising a flag. I fixed it with a new commit.

ipopescu commented on glossary/C.rst:
```diff
@@ -0,0 +1,73 @@
+
+CSPR
+^^^^
+CSPR is the Casper token pre-defined on the Casper network and used to pay for transaction execution.
```
Updated the **CSPR** entry and added an entry for **Staking**.

**CSPR**
CSPR is the Casper token pre-defined on the Casper network and used to pay for transaction execution and for `staking <S.html#staking>`_ (securing the network).

**Staking**
A feature of Proof-of-Stake protocols that allows token holders to actively participate in the protocol, thus securing the network. The `Staking Guide <https://docs.casperlabs.io/en/latest/staking/index.html>`_ highlights the steps required to stake the CSPR token on the Casper network.

ipopescu commented on glossary/E.rst:
```diff
@@ -0,0 +1,22 @@
+
+Eviction
+^^^^^^^^
+Deactivate validators who fail to participate in an era, disabling their bid and suspending their participation until they signal readiness to resume participation by invoking a particular entry point in the auction contract.
```
Updated the Eviction definition. I also defined a new term  "auction contract", in addition to the auction.

Auction
^^^^^^^
The auction determines the composition of the validator set for each era of the protocol. It is a "first-price" auction (where winning bids become stakes) with a fixed number of spots, chosen to balance security with performance. Because rewards are proportional to the stake, it is expected that this competitive mechanism will provide a powerful impetus for staking as many tokens as possible.

Auction contract
^^^^^^^^^^^^^^^^
The auction contract acts as a front-end user interface to the `auction <auction>`_ by accepting bids from validators and delegators directly. It also contains the logic necessary for carrying out the auction.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,91 @@
+
+Booking block
+^^^^^^^^^^^^^
+The first block that is less than 11 days way from the beginning of an era. It defines the new era's validator set, weights, but not yet the leader schedule. Each block between the booking block and the key block contains one random bit that contributes to the seed used for the pseudorandom leader schedule.
```
Thanks, @afck & @MParlikar. I've updated the _booking block_ definition and added a new definition for the _auction delay._ 

**Booking block**
The first block in an era, which considers the `auction delay <A.html#auction-delay>`_. In other words, the last block of era N is the booking block for era N + `auction delay <A.html#auction-delay>`_ + 1, where the auction delay is the number of full eras that pass between the booking block and the era whose validator set it defines. The auction delay is configurable and can be several eras.

**Auction delay**
The number of full eras that pass between the `booking block <B.html#booking-block>`_ and the era whose validator set it defines. The auction delay is configurable and can be several eras.

ipopescu commented:
Update on 2/26
-  Addressed review feedback.
- Updated the **gas**, **cspr**, **booking block**, and **eviction** definitions, and added new entries for: **auction contract**, **auction delay**, **staking**, and **operator**.
- Fixed the sorting order in file B and a few other entries.
- Removed a few duplicates.
- Combed through everything one more time and ran Grammarly.
- Checked for any statements that did not sound configurable (such as 11 days in the booking block).

afck commented on glossary/S.rst:
```diff
@@ -0,0 +1,38 @@
+^^^^^^^^^^^^
+The *session code* is a field contained in a deployment directive. The *session code* contains the code the user wishes to execute against the blockchain.  When the session code executes, it performs a transaction. 
+
+Shard
+^^^^^
+A shard is a subset of the blocks and transactions that have to be processed by only a specific subset of validators.
+
```
Should we remove that? We don't currently do sharding.

ipopescu replied:
Removed

afck commented on glossary/S.rst:
```diff
@@ -0,0 +1,38 @@
+
+Slashing
+^^^^^^^^
+In Proof-of-Stake, the deposit acts as collateral. The validator guarantees that it correctly follows the protocol. If the validator node creates an equivocation or fails to send messages for too long, the deposited amount gets "slashed," i.e., a part of it is removed.
```
```suggestion
In Proof-of-Stake, the deposit acts as collateral. The validator guarantees that it correctly follows the protocol. If the validator node violates the protocol, the deposited amount gets "slashed," i.e., removed.
```

ipopescu replied:
Updated

ashok-cl commented on glossary/N.rst:
```diff
@@ -0,0 +1,32 @@
+
+Node operator
+^^^^^^^^^^^^^
+Anyone who wishes to run node infrastructure on the Casper network either as a standalone private network, or as part of the public network. Synonymous to a `validator <V.html#validator>`_.
```
@MParlikar - not sure we need to or should specifically mention 'private network' here.

ipopescu replied:
Removed the mention of private network and added a new definition for Operator based on Ed's suggestion.

afck commented on glossary/S.rst:
```diff
@@ -0,0 +1,38 @@
+
+Staker
+^^^^^^
+A person that deposits tokens in the `proof-of-stake <P.html#proof-of-stake>`_ contract. Sometimes a staker is also a `validator <V.html#validator>`_ or a `delegator <D.html#delegator>`_. Stakers take on the slashing risk in exchange for rewards. Stakers will deposit their `tokens <T.html#token>`_ by sending a bonding request in the form of a transaction (deployment) to the system. If a validator ...
```
Not sure whether we still have a (user-visible) distinction of "proof of stake contract" vs. "auction contract", or we should just always use the term "auction contract"?

ipopescu replied:
I see the following explanation in the Execution Semantics: "Payment code ultimately provides its payment by performing a token transfer into the proof-of-stake contract’s payment purse. If payment is not given or not enough is transferred, then payment execution is not considered successful." 
https://docs.casperlabs.io/en/latest/implementation/execution-semantics.html?highlight=payment%20code#payment-code
I will keep the "proof of stake contract" until I hear otherwise.

afck commented on glossary/T.rst:
```diff
@@ -0,0 +1,52 @@
+
+Testnet
+^^^^^^^
+The feature-complete, decentralized public Casper platform for testing.
```
Do we still need entries for "devnet" and "testnet" — or should we have one for "delta" now instead?

ipopescu replied:
For the sake of simplicity and since we are close to mainnet, I will remove devnet and testnet. I am not sure about Delta, I will find out.

ipopescu replied:
Let's remove Delta and focus on mainnet.

afck commented on glossary/T.rst:
```diff
@@ -0,0 +1,52 @@
+*  **Payment code** – [mandatory] the "short-leash" code (=WASM program) which will pay the execution fee
+*  **Session code** – [mandatory] the code (= WASM program) which the user wishes to execute against the blockchain
+*  **Gas price** – [mandatory] price of the gas unit (expressed as a number of tokens per single gas unit)
+*  **Nonce** – [mandatory] the number which prevents replay attacks by making each signature valid for a single execution
```
Obsolete?
(But maybe other fields where added?)

EdHastingsCasperLabs replied:
Nonce is definitely obsolete...it was even obsolete in the scala hybrid era.

ipopescu replied:
Removed "nonce". I still need to find out if the other fields are correct.

ipopescu replied:
Pointed Transaction to Deploy, and kept the simpler Deploy definition.

afck commented on glossary/T.rst:
```diff
@@ -0,0 +1,52 @@
+*  **Gas price** – [mandatory] price of the gas unit (expressed as a number of tokens per single gas unit)
+*  **Nonce** – [mandatory] the number which prevents replay attacks by making each signature valid for a single execution
+*  **TTL** - [optional] upper limit for P-time of a block this transaction may be included in
+*  **Block hook** - [optional] semantic dependency of this transaction, expressed as a hash of a block that is 
```
Are dependencies specified as _deploy_ hashes instead of _block_ hashes now?

ipopescu replied:
Removed these details.

afck commented on glossary/T.rst:
```diff
@@ -0,0 +1,52 @@
+*  **Nonce** – [mandatory] the number which prevents replay attacks by making each signature valid for a single execution
+*  **TTL** - [optional] upper limit for P-time of a block this transaction may be included in
+*  **Block hook** - [optional] semantic dependency of this transaction, expressed as a hash of a block that is 
+*  **Signature** – [mandatory] must be done via a private or secret key (applied over the payment code, session code, conversion rate, or nonce), to ensure the correct execution of intended operations
```
```suggestion
*  **Signature** – [mandatory] must be done via a secret key (applied over the payment code, session code, conversion rate, or nonce), to ensure the correct execution of intended operations
```
(I think it's less confusing to consistently use "secret key". Otherwise it looks as if those were distinct keys.)

afck commented on glossary/V.rst:
```diff
@@ -0,0 +1,18 @@
+* `block validation <B.html#block-validation>`_
+* `block gossiping <B.html#block-gossiping>`_
+
+Validators are bonded because they are responsible for progressing the system's state as clients use it (e.g., sending deploys). Validators and `stakers <S.html#staker>`_ can lose their bond (be slashed) for not following the protocol correctly. Validators are also paid for by creating blocks (also by validating blocks – though this is only indirectly; validators cannot be paid for if they do not ...
```
Are they still synonymous? Maybe we use both terms in multiple, slightly different meanings sometimes. :confused:

ipopescu replied:
You're right, one can operate a node without being a validator. I will take this out.

EdHastingsCasperLabs replied:
Any one running a node is an _operator_ 

An operator that has staked a bid but does not currently have a validator slot is a _staked operator_

An operator who's bid has won a validator slot in the auction for a specific era is a _validator_ in that era. 

It is common for a staked operator to win an auction slot for many eras in a row, and thus be a validator for a range of eras.

ipopescu replied:
Thanks @EdHastingsCasperLabs! I added a new entry for _Operator_ with these definitions. I pointed "node operator" to "operator".

ashok-cl commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Participate in consensus
+^^^^^^^^^^^^^^^^^^^^^^^^
+The process of following the `consensus <C.html#consensus>`_ algorithm. The primary participants are `validators <V.html#validator>`_ and `delegators <D.html#delegator>`_.
```
_The primary participants are `validators <V.html#validator>`_ and `delegators <D.html#delegator>`_._

Maybe more accurate to say 'the primary participants are validators that are bonded with their stake and are part of the validator set for that particular era. Delegators participate indirectly by delegating their tokens to one or more of these validators and contribute by increasing the total stake that goes towards ensuring security of the network' @afck

afck commented on implementation/p2p.rst:
```diff
@@ -303,7 +303,7 @@ following parameters:
    notified about, but if multiple iterations are needed to find the connection
    points then they can be further back the DAG.
--  ``known_block_hashes`` can be supplied by the caller to provide an early exist
+-  ``known_block_hashes`` can be supplied by the caller to provide an early exit
```
At a glance, it looks like most of this file is outdated (refers to the old Scala code).

ipopescu replied:
This file will be updated as part of https://github.com/CasperLabs/docs/pull/133.

afck reviewed (commented):
Looks great! :+1: 

Please take my comments only as _suggestions_; I'm not sure what exactly should be the scope of the glossary, and whether everyone agrees with what I wrote. In particular, someone more familiar with the contract runtime should review this as well.

ashok-cl commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Primary token
+^^^^^^^^^^^^^
+The Casper platform defines only one token which one can pay for transaction execution. Presently this is called `CSPR <C.html#cspr>`_.
```
The Casper platform defines only one token with which one can pay for transaction execution and the same is also used for staking to secure the network and earn rewards.

ipopescu replied:
I decided to point this directly CSPR by saying "See CSPR."

ashok-cl commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Proto block
+^^^^^^^^^^^
+The block proposed by the leader, which the consensus processes (in `highway <H.html#highway>`_). Only after consensus is complete, the proto block is executed and the global state is updated.
```
Maybe add: 
A leader is selected from the validator set of that era for each round and the chance of getting selected as a leader is in proportion to their stake in that era.

ipopescu replied:
Added.

ashok-cl commented on glossary/S.rst:
```diff
@@ -0,0 +1,34 @@
+
+Slashing
+^^^^^^^^
+In Proof-of-Stake, the deposit acts as collateral. The validator guarantees that it correctly follows the protocol. If the validator node creates an equivocation or fails to send messages for too long, the deposited amount gets "slashed," i.e., a part of it is removed.
```
_fails to send messages for too long_ - This is not true. They only get evicted and don't receive any rewards. @afck

ipopescu replied:
Updated to: "The validator guarantees that it correctly follows the protocol. If the validator node violates the protocol, the deposited amount gets "slashed," i.e., a part of it is removed."

afck commented on glossary/N.rst:
```diff
@@ -0,0 +1,32 @@
+
+You will encounter different types of nodes on the network:
+
+* **Bonded node**: any node that can be trusted on the network.
```
That's true, "can be trusted" is a bit vague.

ipopescu replied:
Updated based on Ashok's suggestion

afck commented on glossary/E.rst:
```diff
@@ -0,0 +1,22 @@
+
+Era
+^^^
+The Highway protocol execution is split into eras, and at the start of each era, every validator is initialized. Validators are forced to store the whole DAG, even the units that were created at the very beginning of the protocol execution. To address the large storage challenge, the protocol execution is divided into eras. In each era, 1000 new blocks are added to the blockchain, and importantly ...
```
Yes, I made a proposal above.
I agree, it's a config setting, and probably not 1000 blocks.

ipopescu replied:
Updated

afck commented on glossary/B.rst:
```diff
@@ -0,0 +1,91 @@
+
+Block finality
+^^^^^^^^^^^^^^
+Block finality in the `Highway <H.html#highway>`_ protocol is expressed by a fraction of nodes that would need to break the protocol rules in order for a block to be reverted. During periods of honest participation finality of blocks might reach well beyond 1/3 (as what would be the maximum for classical protocols), up to even 1 (where 1 is complete certainty).
```
I made a proposal above.
Yes, a finalized block's fault tolerance can go higher than 1/3.

ipopescu replied:
Updated

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+
+* **J-time**: sum of gas burned in J-past-cone of B (excluding transactions in B)
+* **P-time**: sum of gas burned in P-past-cone of B (excluding transactions in B)
+
```
Removed with commit 69fbce78c802ed8ee5691e640344202208ab8ae5.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+^^^^^
+Used in two contexts:
+
+#. A data structure containing a collection of transactions. Blocks form the BlockDAG and, as such, are a primary structure of the blockchain.
```
@afck removed block DAG from all explanations with commit 832818c6bf4efbc40a251d41f077c5a97ec2b65a.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+
+Each block has a globally unique ID, achieved by hashing the contents of the block.
+
+Each block points to his parents (one or many) and justification blocks (one or many). Parents are a subset of justifications. An exception is the Genesis block, which has no parents and no justifications.
```
Updated with commit cc6a6f5f372d8f27f8b0578873caf2babfb2a5ad. Just fyi, I couldn't apply the commit directly (due to a prior merge from master), but I appreciate this format!

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+Block creation
+^^^^^^^^^^^^^^
+Block creation is the process of computing the deployment results (not contained in a block known by the node) and collecting the results that belong together into a block. The order in which deployments are applied to the global state does not change the outcome.
+Block creation is always followed by "block proposal" (see below) because it does not make sense to create a block without sending it out to the network. Note that only validators can create valid blocks.
```
@afck submitted an updated definition for block creation with commit 525219238d13c35510c5b756aa9eb2297842c085. I also added 2 related definitions for **root hash** and **Merkle tree** with this commit.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+* Signature is from a bonded participant
+* Choice of parent block(s) follow from justifications (fork-choice correctness)
+* Deploys' update to VM state is as reported by the block (see "block processing")
+* Bonds map matches the PoS contract state
```
Agreed. Keeping simple is better in a glossary. Will remove the list.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+
+Booking block
+^^^^^^^^^^^^^
+The first block in an era, which considers the `auction delay <A.html#auction-delay>`_. In other words, the last block of era N is the booking block for era N + `auction delay <A.html#auction-delay>`_ + 1, where the auction delay is the number of full eras that pass between the booking block and the era whose validator set it defines. The auction delay is configurable and can be several eras long ...
```
Incorporated the suggestion

ipopescu commented on glossary/C.rst:
```diff
@@ -0,0 +1,73 @@
+
+Consensus
+^^^^^^^^^
+An algorithm used to mandate agreement on the `blockDAG <B.html#blockdag>`_ between all nodes. The blockDAG, although being built in a decentralized way, eventually converges so that all nodes eventually agree on whether a given block B is part of the blockDAG or not, and what is the `J-past-cone <C.html#cone>`_ and `P-past-cone <C.html#cone>`_ of B.
```
Updated to include suggestion

ipopescu commented on glossary/D.rst:
```diff
@@ -0,0 +1,37 @@
+^^^^^^^^^^^^^^^^^^^^^^
+A directed acyclic graph is a directed graph that has no cycles. A vertex *v* of a directed graph is said to be reachable from another vertex *u* when there exists a path that starts at *u* and ends at *v*. 
+
+In the Highway protocol, validators exchange messages in order to reach consensus on proposed blocks and hence validate one of many branches of the produced blockchain. As a way of capturing and spreading the different validators’ knowledge about the already existing messages, it adopts the DAG framework, in which every message broadcast by a validator refers to a certain set of messages sent by  ...
```
Agreed. Let's keep the glossary simple. Removed this section.

ipopescu commented on glossary/E.rst:
```diff
@@ -0,0 +1,22 @@
+
+Era
+^^^
+The Highway protocol execution is split into eras, and at the start of each era, every validator is initialized. Validators are forced to store the whole DAG, even the units created at the beginning of the protocol execution. The protocol execution is divided into eras to address the large storage required. In each era, 1000 new blocks are added to the blockchain, and importantly a new instance o ...
```
Incorporated the suggestion. Yes, we can mention the era length when it's decided.

ipopescu commented on glossary/F.rst:
```diff
@@ -0,0 +1,14 @@
+Finality
+^^^^^^^^
+See `block finality <B.html#block-finality>`_.
+
+Finality theorem
+^^^^^^^^^^^^^^^^
+In the Highway protocol, if an honest validator reaches finality with confidence threshold t >= f for a given valid block B, then no honest validator will ever reach finality with confidence threshold t for a block competing with B.
```
Removed

ipopescu commented on glossary/C.rst:
```diff
@@ -0,0 +1,73 @@
+
+CSPR
+^^^^
+CSPR is the Casper token pre-defined on the Casper network and used to pay for transaction execution and for `staking <S.html#staking>`_ (securing the network).
```
Included. "The total number of CSPR tokens is 10 billion."

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+
+Block finality
+^^^^^^^^^^^^^^
+Block finality in the `Highway <H.html#highway>`_ protocol is expressed by a fraction of nodes that would need to break the protocol rules in order for a block to be reverted. During periods of honest participation finality of blocks might reach well beyond 1/3 (as what would be the maximum for classical protocols), up to even 1 (where 1 is complete certainty).
```
Feedback incorporated

ipopescu commented on glossary/G.rst:
```diff
@@ -0,0 +1,24 @@
+
+Genesis
+^^^^^^^
+The first `block <B.html#block>`_ in the `blockchain <B.html#blockchain>`_. P-future-chain of Genesis includes all blocks ever created.
```
Incorporated this suggestion

ipopescu commented on glossary/G.rst:
```diff
@@ -0,0 +1,24 @@
+
+Technically, a `global state <G.html#global-state>`_ is a (possibly extensive) collection of key-value pairs, where the keys are references (Refs), and the values are large binary objects (BLOBs).
+
+For every `block <B.html#block>`_ B in the `blockDAG <B.html#blockdag>`_, one can compute the `global state <G.html#global-state>`_ achieved by executing all `transactions <T.html#transaction>`_ in this block and its parent blocks (recursively, using the past-cone path). This global state is a global state, as seen by block B.
```
Feedback incorporated

ipopescu commented on glossary/I.rst:
```diff
@@ -0,0 +1,9 @@
+`A <A.html>`_  `B <B.html>`_  `C <C.html>`_  `D <D.html>`_  `E <E.html>`_  `F <F.html>`_  `G <G.html>`_  `H <H.html>`_  `I <I.html>`_  `J <J.html>`_  `K <K.html>`_  `L <L.html>`_  `M <M.html>`_  `N <N.html>`_  `O <O.html>`_  `P <P.html>`_  `Q <Q.html>`_  `R <R.html>`_  `S <S.html>`_  `T <T.html>`_  `U <U.html>`_  `V <V.html>`_  `W <W.html>`_  `X <X.html>`_  `Y <Y.html>`_  `Z <Z.html>`_  
+============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== ============== 
+
+Coming soon
+^^^^^^^^^^^
```
Let's make it empty, and keep all the letters for easier navigation. Removed "coming soon" from all empty entries.

ipopescu commented on glossary/N.rst:
```diff
@@ -0,0 +1,32 @@
+#. COMM - contains the logic for message passing between nodes on the network (via TCP)
+#. Consensus - contains the logic to update the `blockstore <B.html#blockstore>`_ and `global state <G.html#global state>`_ so that agreement is maintained between nodes.
+#. Execution Engine - this is where the actual `transaction <T.html#transaction>`_ execution happens in a WASM interpreter. It contains the logic to record updates to the global state that transactions will do, as well as to apply those updates under the direction of consensus.
+#. Storage - this is where the blockstore and the global state are stored.
```
Removed to keep it clean and simple.

ipopescu commented on glossary/N.rst:
```diff
@@ -0,0 +1,32 @@
+
+Nonce
+^^^^^
+The number that prevents replay attacks by making each signature valid for a single execution.
```
Removed nonce and will follow up on TTL and dependencies.

ipopescu commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Participate in consensus
+^^^^^^^^^^^^^^^^^^^^^^^^
+The process of following the `consensus <C.html#consensus>`_ algorithm. The primary participants are `validators <V.html#validator>`_ and `delegators <D.html#delegator>`_.
```
Removed delegators. Kept validators as primary consensus participants.

ipopescu commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Primary token
+^^^^^^^^^^^^^
+The Casper platform defines only one token which one can pay for transaction execution. Presently this is called `CSPR <C.html#cspr>`_.
```
Updated to "with which one can pay".

ipopescu commented on glossary/P.rst:
```diff
@@ -0,0 +1,42 @@
+
+Proof of stake
+^^^^^^^^^^^^^^
+Proof of stake (PoS) is a type of consensus mechanism by which a cryptocurrency blockchain network achieves distributed consensus. The voting power is proportional to the number of tokens (digital currency specific to this system). A popular choice in such systems is to periodically delegate a fixed size committee of participants, which then is responsible for running the consensus on which block ...
```
Agreed. I have seen Proof-of-Stake more often and will use that. I will adopt your explanation and put it up for a final review.

ipopescu commented on glossary/R.rst:
```diff
@@ -0,0 +1,22 @@
+
+Reward
+^^^^^^
+Validators receive rewards for participating in consensus by voting on blocks sending finality signatures (finalizing blocks). There is no precise per-block reward. Rewards are split proportionally among stakes within an era. If a validator is offline or cannot vote on many blocks, the rewards earned are also reduced. Delegators can only receive a proportional amount of the validator's rewards mi ...
```
Updated

ipopescu commented on glossary/R.rst:
```diff
@@ -0,0 +1,22 @@
+
+Runtime
+^^^^^^^
+The Casper runtime layer encompasses deployment and execution of `smart contracts <S.html#smart-contracts>`_, `session code <S.html#session-code>`_, and other activity that performs computation on the `global state <G.html#global-state>`_. 
```
Removed. Renamed _Casper runtime_ to _Contract runtime_. I will take the second suggestion in a separate ticket.

ipopescu commented:
Addressed most feedback. I have a few open questions which I will summarise below and follow up.

What are the current mandatory fields in a transaction?
https://github.com/CasperLabs/docs/pull/124#discussion_r583822415 

Do we use “block store” or “block storage”?
https://github.com/CasperLabs/docs/pull/124#discussion_r583788402 

Definition of a Deploy TTL and a Dependency? Coming from this comment:
https://github.com/CasperLabs/docs/pull/124#discussion_r583811743 

Are dependencies specified as deploy hashes instead of block hashes now?
https://github.com/CasperLabs/docs/pull/124#discussion_r583822660 

Do we still have a (user-visible) distinction of "proof of stake contract" vs. "auction contract", or we should just always use the term "auction contract"?
https://github.com/CasperLabs/docs/pull/124#discussion_r583821277 

Do we want a definition for Delta in the glossary?
https://github.com/CasperLabs/docs/pull/124#discussion_r586508208 

Not sure we need to or should specifically mention 'private network' in the Node operator definition: anyone who wishes to run node infrastructure on the Casper network either as a standalone private network, or as part of the public network.
https://github.com/CasperLabs/docs/pull/124#discussion_r583820552 

If anyone disagrees with Andreas’ controversial opinion regarding Proof-of-Stake, let me know. Otherwise I will go with his definition.
https://github.com/CasperLabs/docs/pull/124#discussion_r583815535

EdHastingsCasperLabs commented on glossary/T.rst:
```diff
@@ -0,0 +1,48 @@
+^^^^^
+A type of cryptocurrency that represents an asset. See `CSPR <C.html#cspr>`_.
+
+Transaction
```
This is a _Deploy_ in our system.

Suggest this should be moved under D, and here under Transaction link to Deploy "See `Deploy <D.html#deploy>"

ipopescu replied:
Done. This also resolves the above comment about the transaction fields.

ipopescu commented on glossary/F.rst:
```diff
@@ -0,0 +1,14 @@
+^^^^^^^^
+See `block finality <B.html#block-finality>`_.
+
+Finality theorem
```
Andreas suggested to remove the finality theorem all-together. Those interested can read the paper.

ipopescu commented on glossary/L.rst:
```diff
@@ -0,0 +1,10 @@
+
+Liveness theorem
+^^^^^^^^^^^^^^^^
+In the Highway protocol, for every confidence threshold 0 <= t < n/3 , if f <= t and c < (n-3t)/2, then the chain of blocks finalized with confidence t grows indefinitely for each honest validator.
```
Removed the _liveness theorem_, and those interested can read the paper. We do cover _liveness_ separately. Discussed this with Andreas.

ipopescu commented on glossary/B.rst:
```diff
@@ -0,0 +1,92 @@
+
+Collectively, the collection of all blocks is called a BlockDAG. Every node keeps a "local" view of the BlockDAG, attempting to capture and include all the updates across the nodes' P2P network. In other words, a node continually tries to catch up with the evolution of the BlockDAG, which is happening when the validators attempt to create new blocks.
+
+Blockstore
```
I see that in the _node operator/setup.rst_ we use block store. I will use this for now until I hear otherwise.

ipopescu commented:
This PR is ready for a final review and hopefully a merge. There are 3 open comments, but maybe we can address them right before the merge.

1. Should we use block store or block storage? (comment https://github.com/CasperLabs/docs/pull/124#discussion_r583788402)
2. Should we have definitions for a deployment TTL and deploy dependencies? (comment https://github.com/CasperLabs/docs/pull/124#discussion_r583811743)
3. Do we still have a (user-visible) distinction of "proof of stake contract" vs. "auction contract", or we should just always use the term "auction contract"? I decided to keep it until I hear otherwise. It's mentioned in the execution semantics. (comment https://github.com/CasperLabs/docs/pull/124#discussion_r583821277)

MParlikar reviewed (approved):
Great job
